File: 0874/Solution.java
class Solution {
    int HASH_MULTIPLIER = 60001;
    public int robotSim(int[] commands, int[][] obstacles) {
        int max = 0;
        int x = 0, y = 0, dir = 0;
        int[][] dirs = new int[][]{{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        Set<Integer> obs = new HashSet<>();

        for(int[] o : obstacles){
           obs.add(hashCoordinates(o[0], o[1]));
        }

        for(int command : commands){
            switch(command){
                case -1 : dir = (dir + 1) % 4; break;
                case -2 : dir = (dir + 3) % 4; break;
                default : {
                    while(command-- > 0){
                        int newX = x + dirs[dir][0];
                        int newY = y + dirs[dir][1];

                        if(obs.contains(hashCoordinates(newX, newY))) break;
                        
                        x = newX; 
                        y = newY;
                    }
                    max = Math.max(max, x * x + y * y);
                }
            }
        }
        return max;
    }

    int hashCoordinates(int x, int y){
        return x + y * HASH_MULTIPLIER;
    }
}
// time O(n + k)
// space O(k)

File: 0874/Solution.js
/**
 * @param {number[]} commands
 * @param {number[][]} obstacles
 * @return {number}
 */
var robotSim = function(commands, obstacles) {
    const HASH_MULTIPLIER = 60001;
    const hashCoordinates = (x, y) => {
        return x + y * HASH_MULTIPLIER;
    };
    const obs = new Set(obstacles.map(obstacle => hashCoordinates(obstacle[0], obstacle[1])));
    let max = 0, x = 0, y = 0, dir = 0;
    let dirs = [[0, 1], [1, 0], [0, -1], [-1, 0]];
    
    for(let command of commands){
        if(command == -1){
            dir = (dir + 1) % 4;
        }else if(command == -2){
            dir = (dir + 3) % 4;
        }else{
            while(command-- > 0){
                let newX = x + dirs[dir][0];
                let newY = y + dirs[dir][1];

                if(obs.has(hashCoordinates(newX, newY)))
                    break;
                
                x = newX, y = newY;
        
            }
        }
        max = Math.max(max, x * x + y * y);
    }
    return max;
};
